File: backend/app/scrapers/hatena.py
"""
はてなブックマークのスクレイピング機能
"""
import httpx
from bs4 import BeautifulSoup
from typing import List, Dict
import logging

logger = logging.getLogger(__name__)

async def fetch_hatena_hot() -> List[Dict]:
    """はてなブックマーク人気エントリーを取得"""
    url = "https://b.hatena.ne.jp/hotentry?mode=rss"

    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            response = await client.get(url)
            response.raise_for_status()
            xml_content = response.text

            return parse_hatena_rss(xml_content)
    except Exception as e:
        logger.error(f"Error fetching Hatena hot entries: {e}")
        raise

async def fetch_hatena_latest() -> List[Dict]:
    """はてなブックマーク新着エントリーを取得"""
    url = "https://b.hatena.ne.jp/entrylist?mode=rss"

    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            response = await client.get(url)
            response.raise_for_status()
            xml_content = response.text

            return parse_hatena_rss(xml_content)
    except Exception as e:
        logger.error(f"Error fetching Hatena latest entries: {e}")
        raise

def parse_hatena_rss(xml_content: str) -> List[Dict]:
    """はてなブックマークRSSをパース"""
    soup = BeautifulSoup(xml_content, 'xml')
    items = soup.find_all('item')

    entries = []
    for item in items:
        try:
            title = item.find('title').text if item.find('title') else ""
            description = item.find('description').text if item.find('description') else ""
            link = item.find('link').text if item.find('link') else ""

            # はてなブックマーク独自の要素
            comments_url = ""
            bookmark_count = 0

            # bookmarkCommentListPageUrl を探す
            comments_url_tag = item.find('bookmarkCommentListPageUrl')
            if comments_url_tag:
                comments_url = comments_url_tag.text

            # bookmarkcount を探す
            bookmark_count_tag = item.find('bookmarkcount')
            if bookmark_count_tag:
                try:
                    bookmark_count = int(bookmark_count_tag.text)
                except ValueError:
                    bookmark_count = 0

            entries.append({
                "title": title,
                "description": description,
                "url": link,
                "comments_url": comments_url,
                "bookmark_count": bookmark_count
            })
        except Exception as e:
            logger.warning(f"Error parsing RSS item: {e}")
            continue

    return entries

async def fetch_hatena_comments(url: str) -> List[Dict]:
    """はてなブックマークのコメントを取得"""
    try:
        async with httpx.AsyncClient(timeout=30.0, follow_redirects=True) as client:
            response = await client.get(url)
            response.raise_for_status()
            html_content = response.text

            return parse_hatena_comments(html_content)
    except Exception as e:
        logger.error(f"Error fetching Hatena comments from {url}: {e}")
        raise

def parse_hatena_comments(html_content: str) -> List[Dict]:
    """はてなブックマークのコメントHTMLをパース"""
    soup = BeautifulSoup(html_content, 'html.parser')

    # esuna_oldと同じセレクターを使用
    comment_nodes = soup.select('.js-bookmarks-recent > .entry-comment-contents')

    comments = []
    for comment_node in comment_nodes:
        try:
            user_name = ""
            text = ""

            # ユーザー名
            user_name_node = comment_node.select_one('.entry-comment-username > a')
            if user_name_node:
                user_name = user_name_node.get_text(strip=True)

            # コメント本文
            text_node = comment_node.select_one('.entry-comment-text')
            if text_node:
                text = text_node.get_text(strip=True)

            if user_name or text:  # 空のコメントは除外
                comments.append({
                    "user_name": user_name,
                    "text": text
                })
        except Exception as e:
            logger.warning(f"Error parsing comment: {e}")
            continue

    return comments

File: backend/src/scrapers/hatena.ts
/**
 * はてなブックマークのスクレイピング機能
 */
import * as cheerio from "cheerio";
import type { HatenaEntry, HatenaComment } from "../types";

/**
 * はてなブックマークRSSをパース
 * cheerio で XML を処理する
 */
// 名前空間付き要素 (`hatena:bookmarkcount` 等) は CSS セレクタの `\\:` エスケープが
// 環境（workerd runtime と node）で解釈差を起こす。子要素を直接走査して tagName で
// 振り分ければ namespace の有無に依存せず取れる。
function findChildText(
  $: cheerio.CheerioAPI,
  $item: cheerio.Cheerio<any>,
  tagName: string,
): string {
  let value = "";
  $item.children().each((_, child) => {
    if (value) return;
    if ((child as { tagName?: string }).tagName === tagName) {
      value = $(child).text();
    }
  });
  return value;
}

function parseHatenaRss(xmlContent: string): HatenaEntry[] {
  const $ = cheerio.load(xmlContent, { xmlMode: true });
  const entries: HatenaEntry[] = [];

  $("item").each((_, item) => {
    try {
      const $item = $(item);
      const title = findChildText($, $item, "title");
      const description = findChildText($, $item, "description");
      const url = findChildText($, $item, "link");

      const commentsUrl = findChildText($, $item, "hatena:bookmarkCommentListPageUrl");
      const bookmarkCountText = findChildText($, $item, "hatena:bookmarkcount");
      const bookmarkCount = bookmarkCountText
        ? parseInt(bookmarkCountText, 10) || 0
        : 0;

      entries.push({
        title,
        description,
        url,
        comments_url: commentsUrl,
        bookmark_count: bookmarkCount,
      });
    } catch (e) {
      console.warn("Error parsing RSS item:", e);
    }
  });

  return entries;
}

/**
 * はてなブックマーク人気エントリーを取得
 */
export async function fetchHatenaHot(): Promise<HatenaEntry[]> {
  const url = "https://b.hatena.ne.jp/hotentry?mode=rss";

  const response = await fetch(url, {
    signal: AbortSignal.timeout(30_000),
  });
  if (!response.ok) {
    throw new Error(`HTTP ${response.status}: ${response.statusText}`);
  }
  const xmlContent = await response.text();
  return parseHatenaRss(xmlContent);
}

/**
 * はてなブックマーク新着エントリーを取得
 */
export async function fetchHatenaLatest(): Promise<HatenaEntry[]> {
  const url = "https://b.hatena.ne.jp/entrylist?mode=rss";

  const response = await fetch(url, {
    signal: AbortSignal.timeout(30_000),
  });
  if (!response.ok) {
    throw new Error(`HTTP ${response.status}: ${response.statusText}`);
  }
  const xmlContent = await response.text();
  return parseHatenaRss(xmlContent);
}

/**
 * はてなブックマークのコメントHTMLをパース
 */
function parseHatenaComments(htmlContent: string): HatenaComment[] {
  const $ = cheerio.load(htmlContent);
  const comments: HatenaComment[] = [];

  $(".js-bookmarks-recent > .entry-comment-contents").each((_, node) => {
    try {
      const $node = $(node);
      const userName =
        $node.find(".entry-comment-username > a").text().trim() || "";
      const text = $node.find(".entry-comment-text").text().trim() || "";

      if (userName || text) {
        comments.push({ user_name: userName, text });
      }
    } catch (e) {
      console.warn("Error parsing comment:", e);
    }
  });

  return comments;
}

/**
 * はてなブックマークのコメントを取得
 */
export async function fetchHatenaComments(
  url: string
): Promise<HatenaComment[]> {
  const response = await fetch(url, {
    signal: AbortSignal.timeout(30_000),
    redirect: "follow",
  });
  if (!response.ok) {
    throw new Error(`HTTP ${response.status}: ${response.statusText}`);
  }
  const htmlContent = await response.text();
  return parseHatenaComments(htmlContent);
}
